File: python/grpc_parser.py
from aptos.util.timestamp import timestamp_pb2
from aptos.transaction.testing1.v1 import transaction_pb2
from create_table import Event
import datetime

# INDEXER_NAME is used to track the latest processed version
INDEXER_NAME = "python_example_indexer"


def parse(transaction: transaction_pb2.Transaction):
    # Custom filtering
    # Here we filter out all transactions that are not of type TRANSACTION_TYPE_USER
    if transaction.type != transaction_pb2.Transaction.TRANSACTION_TYPE_USER:
        return

    # Parse Transaction struct
    transaction_version = transaction.version
    transaction_block_height = transaction.block_height
    inserted_at = parse_timestamp(transaction.timestamp)
    user_transaction = transaction.user

    # Parse Event struct
    event_db_objs: list[Event] = []
    for event_index, event in enumerate(user_transaction.events):
        creation_number = event.key.creation_number
        sequence_number = event.sequence_number
        account_address = standardize_address(event.key.account_address)
        type = event.type_str
        data = event.data

        # Create an instance of Event
        event_db_obj = Event(
            creation_number=creation_number,
            sequence_number=sequence_number,
            account_address=account_address,
            transaction_version=transaction_version,
            transaction_block_height=transaction_block_height,
            type=type,
            data=data,
            inserted_at=inserted_at,
            event_index=event_index,
        )
        event_db_objs.append(event_db_obj)

    return event_db_objs


def parse_timestamp(timestamp: timestamp_pb2.Timestamp):
    datetime_obj = datetime.datetime.fromtimestamp(
        timestamp.seconds + timestamp.nanos * 1e-9
    )
    return datetime_obj.strftime("%Y-%m-%d %H:%M:%S.%f")


def standardize_address(address: str):
    return "0x" + address

File: typescript/grpc_parser.ts
import {
  DeleteModule,
  Transaction,
  WriteSetChange,
} from "./aptos/transaction/testing1/v1/transaction_pb";

export function parse(transaction: Transaction): void {
  // Custom filtering: Filter out all transactions that are not User Transactions
  if (
    transaction.getType() != Transaction.TransactionType.TRANSACTION_TYPE_USER
  ) {
    return;
  }

  // Parse Transaction object
  const transactionVersion = transaction.getVersion();
  const transactionBlockHeight = transaction.getBlockHeight();
  const transactionInfo = transaction.getInfo();

  if (transactionInfo == null) {
    return;
  }

  const writeSetChanges = transactionInfo.getChangesList();
  const parsedWriteSetChanges = [];

  // Parse WriteSetChanges
  writeSetChanges.forEach(function (writeSetChange, index) {
    const type = writeSetChange.getType();
    let address = "";
    let stateKeyHash = "";

    switch (type) {
      case WriteSetChange.Type.TYPE_DELETE_MODULE:
      case WriteSetChange.Type.TYPE_DELETE_RESOURCE:
      case WriteSetChange.Type.TYPE_WRITE_MODULE:
      case WriteSetChange.Type.TYPE_WRITE_RESOURCE:
        const change =
          writeSetChange.getDeleteModule() ||
          writeSetChange.getDeleteResource() ||
          writeSetChange.getWriteModule() ||
          writeSetChange.getWriteResource();

        if (change == null) {
          return;
        }

        address = change.getAddress();
        stateKeyHash = change.getStateKeyHash_asB64();
        break;
      case WriteSetChange.Type.TYPE_DELETE_TABLE_ITEM:
      case WriteSetChange.Type.TYPE_WRITE_TABLE_ITEM:
        const tableItemChange =
          writeSetChange.getDeleteTableItem() ||
          writeSetChange.getWriteTableItem();

        if (tableItemChange == null) {
          return;
        }
        stateKeyHash = tableItemChange.getStateKeyHash_asB64();
        break;
      case WriteSetChange.Type.TYPE_UNSPECIFIED:
        return;
    }
    console.log(
      transactionVersion,
      transactionBlockHeight,
      type,
      address,
      stateKeyHash,
      index
    );
  });
}
